sdks/python/apache_beam/runners/interactive/utils_test.py

```diff
@@ -182,6 +183,9 @@ def test_child_module_logger_can_override_logging_level(self, mock_emit):
 @unittest.skipIf(
     not ie.current_env().is_interactive_ready,
     '[interactive] dependency is not installed.')
+@pytest.mark.skipif(
```

Comment:
Why do we need to skip this test (as opposed to failing) if [interactive] is not installed? I'm worried that if the build changes and [interactive] is never installed, we will silently drop test coverage.

Reply:
The tests are skipped for tox environments that does not have interactive dependencies installed. Similar to [gcp], [aws] and etc. Because not all test suites and their test environment has/need all the dependencies installed. These are intentionally skipped.

My worry is that there might be some rare cases that some dependency is not installed correctly and pytest does not skip the test but error out.

I also just mocked out the ie.current_env to gain more stability. It seems that it is the root cause of the flakiness.

Reply:
>In such cases, I would rather the test failed with a descriptive error message. If dependencies are not installed >correctly, that indicates a real problem with either our test setup or the actual code, and we should be aware of it. >But I'm not sure what is the best way to differentiate between intentional skips (as described above) and accidental >skips.

Make sense, let me change the logging to error level. In that case, if the test is not skipped but later somehow the dependency is not installed (or removed), it will error out and expose the problem.


>How do you know?

Based on the stacktrace in the reported JIRA ticket. The actual call is

```
actual = [call('\n            <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.4.1/css/bootstrap.min...">You have limited Interactive Beam features since your ipython kernel is not connected any notebook frontend.</div>')]
```
This indicates `is_in_ipython and not is_in_notebook`, which contradicts with the setup in the test. It means the global instance `ie.current_env()` is modified during the test by other tests. By mocking the IPython environment and the usage of the global  instance, this test is isolated from potential race situations.